### PR TITLE
[WIP] Feature/change database

### PR DESCRIPTION
**Pull request by baumannsven**
Use database abstraction layer instead contao database

### CODE REVIEW

File: src/Data/DefaultDataProvider.php [hunk 1]
```diff
@@ -176,40 +177,67 @@ public function enableDefaultUuidGenerator()
...
+        }
+
+        if (isset($config['connection'])) {
+            if ($config['connection'] instanceof Database) {
+                @trigger_error(
+                    'Contao database is deprecated use doctrine dbal connection. Fallback will be dropped.',
+                    E_USER_DEPRECATED
+                );
+
+                $config['connection'] = System::getContainer()->get('database_connection');
```
Comment (dmolineus) [inline src/Data/DefaultDataProvider.php]: I think that's wrong to use the default connection here. If someone passes a custom connection it's probably not the default one. Maybe you could fetch the connection using reflection from the database object. 

You may have a look at MM 2.1, @discordier does it there somewhere...

File: src/Data/DefaultDataProvider.php [hunk 2]
```diff
@@ -548,157 +557,176 @@ protected function updateModelInDatabase($model)
     {
         $data = $this->convertModelToDataPropertyArray($model);
 
-        $this->objDatabase
-            ->prepare(sprintf('UPDATE %s %%s WHERE id=?', $this->strSource))
-            ->set($data)
-            ->execute($model->getId());
+        $this->connection->update($this->source, $data, [['id' => $model->getId()]]);
```
Comment (dmolineus) [inline src/Data/DefaultDataProvider.php]: Are you sure that the double array syntax is correct here

File: src/Data/TableRowsAsRecordsDataProvider.php [hunk 3]
```diff
@@ -268,47 +268,47 @@ public function save(ModelInterface $objItem, $recursive = false)
             }
 
             if ($intId > 0) {
-                $this->objDatabase
-                    ->prepare(sprintf('UPDATE %s %%s WHERE id=? AND %s=?', $this->strSource, $this->strGroupCol))
+                $this->connection
+                    ->prepare(sprintf('UPDATE %s %%s WHERE id=? AND %s=?', $this->source, $this->strGroupCol))
                     ->set($arrSQL)
```
Comment (dmolineus) [inline src/Data/TableRowsAsRecordsDataProvider.php]: This shouldn't work. Afaik there's no set method using doctrine dbal. The `%s` is also wrong.

File: src/Data/DefaultDataProviderDBalUtils.php [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+
+namespace ContaoCommunityAlliance\DcGeneral\Data;
+
+use ContaoCommunityAlliance\DcGeneral\Exception\DcGeneralRuntimeException;
+use Doctrine\DBAL\Query\QueryBuilder;
+
+/**
+ * Static helper class for DefaultDataProvider.
+ */
+class DefaultDataProviderDBalUtils
```
Comment (dmolineus) [inline src/Data/DefaultDataProviderDBalUtils.php]: Would be great to have this class unit tested.

File: src/Data/TableRowsAsRecordsDataProvider.php [hunk 5]
```diff
@@ -268,47 +268,47 @@ public function save(ModelInterface $objItem, $recursive = false)
...
                 $arrKeep[] = $intId;
             } else {
                 // Force group col value.
-                $arrSQL[$this->strGroupCol] = $objItem->getId();
-                $arrKeep[]                  = $this->objDatabase
-                    ->prepare(sprintf('INSERT INTO %s %%s', $this->strSource))
+                $arrSQL[$this->strGroupCol] = $item->getId();
+                $arrKeep[]                  = $this->connection
+                    ->prepare(sprintf('INSERT INTO %s %%s', $this->source))
                     ->set($arrSQL)
```
Comment (dmolineus) [inline src/Data/TableRowsAsRecordsDataProvider.php]: This shouldn't work. Afaik there's no set method using doctrine dbal. The `%s` is also wrong.

File: src/Data/TableRowsAsRecordsDataProvider.php [hunk 6]
```diff
@@ -268,47 +268,47 @@ public function save(ModelInterface $objItem, $recursive = false)
...
                     ->execute()
                     ->insertId;
```
Comment (dmolineus) [inline src/Data/TableRowsAsRecordsDataProvider.php]: There is no insertId. `$this->connection->lastInsertId()` is the way to go.

File: src/Data/DefaultDataProvider.php (hunk 2 above)
Comment (baumannsven) [reply src/Data/DefaultDataProvider.php]: You're right about that. I'm changing that.

File: src/Data/DefaultDataProvider.php (hunk 1 above)
Comment (discordier) [reply src/Data/DefaultDataProvider.php]: Something along the lines of:
```php
        // BC layer - we used to accept a Contao database instance here.
        if ($dataBase instanceof Database) {
            @trigger_error(
                '"' . __METHOD__ . '" now accepts doctrine instances - ' .
                'passing Contao database instances is deprecated.',
                E_USER_DEPRECATED
            );
            $reflection = new \ReflectionProperty(Database::class, 'resConnection');
            $reflection->setAccessible(true);

            $dataBase = $reflection->getValue($dataBase);
        }
        if (null === $dataBase) {
            @trigger_error(
                'You should pass a doctrine database connection to "' . __METHOD__ . '".',
                E_USER_DEPRECATED
            );
            $dataBase = System::getContainer()->get('database_connection');
        }
```

File: src/Data/TableRowsAsRecordsDataProvider.php (hunk 3 above)
Comment (baumannsven) [reply src/Data/TableRowsAsRecordsDataProvider.php]: Has been fixed

File: src/Data/TableRowsAsRecordsDataProvider.php (hunk 5 above)
Comment (baumannsven) [reply src/Data/TableRowsAsRecordsDataProvider.php]: Has been fixed

File: src/Data/TableRowsAsRecordsDataProvider.php (hunk 6 above)
Comment (baumannsven) [reply src/Data/TableRowsAsRecordsDataProvider.php]: Has been fixed

File: src/Data/DefaultDataProvider.php (hunk 1 above)
Comment (baumannsven) [reply src/Data/DefaultDataProvider.php]: Has been change

File: src/Data/DatabaseUuidIdGenerator.php [hunk 7]
```diff
@@ -50,7 +52,7 @@ public function __construct(\Database $database)
      */
     public function generate()
     {
-        return $this->database->query('SELECT UUID() as id')->first()->id;
+        return $this->connection->query('SELECT UUID() as id')->fetch(\PDO::FETCH_OBJ)->id;
```
Comment (discordier) [inline src/Data/DatabaseUuidIdGenerator.php]: Could be  `return $this->connection->query('SELECT UUID() as id')->fetch(\PDO::FETCH_COLUMN)`?

Comment (baumannsven) [reply src/Data/DatabaseUuidIdGenerator.php]: Has been change.

I have change this in https://github.com/contao-community-alliance/dc-general/pull/386/commits/12ba45fe49de01e41b02cb6ada1f1ef1a6e7a9a0#diff-409a5405b81668016fbcbc5c8a4ae196R461 and https://github.com/contao-community-alliance/dc-general/pull/386/commits/12ba45fe49de01e41b02cb6ada1f1ef1a6e7a9a0#diff-409a5405b81668016fbcbc5c8a4ae196R717 too.

**Comment by baumannsven** [comment]
Type safety handle in contao:
https://github.com/contao/core-bundle/blob/master/src/Resources/contao/library/Contao/Widget.php#L1486

**Comment by baumannsven** [comment]
@discordier I have rebase this branch. Can test this branch for "Type safety".

**Comment by baumannsven** [comment]
Merged in df2f1041

**Pull request closed without merging**